File: day2/main.py
ops = {
    1: lambda x, y: x+y,
    2: lambda x, y: x*y
}

def int_code(code):
    i=0
    while(i+4<len(code)):
        if(code[i] in ops.keys()):
            code[code[i+3]] = ops[code[i]](code[code[i+1]],code[code[i+2]])
            i+=4
        else:
            i+=1
    return code

def read_input(path):
    return [int(x) for x in  open('input.txt').read().split(',')]

if __name__ == "__main__":
    #Read the input
    content = read_input("input.txt")
    #Solve p1
    ##Follow problem instructions to replace values
    content_p1 = content.copy()
    content_p1[1]=12
    content_p1[2]=2
    result = int_code(content_p1)
    print("Result for p1: {}".format(result[0])) 
    #Solve p2 brute force
    for noun in range(99):
        for verb in range(99):
            content_p2 = content.copy()
            content_p2[1]=noun
            content_p2[2]=verb
            r = int_code(content_p2)
            if(r[0]==19690720):
                print("Noun:{}  Verb:{}  Result for p2: {}".format(noun,verb,100*noun+verb))

File: day2/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func main() {
	content := readFile("input.txt")

	var program []int
	for _, value := range strings.Split(content, ",") {
		program = append(program, toInt(value))
	}

	// Solve part1
	program_p1 := make([]int, len(program))
	copy(program_p1, program)
	program_p1[1]=12
	program_p1[2]=2	
	var content_p1 []int = int_code(program_p1)
	{
		fmt.Println("--- Part One ---")
		fmt.Println(content_p1[0])
	}
	//Solve part2
	for noun := 0; noun < 100; noun++ {
		for verb := 0; verb < 100; verb++ {
			program_p2 := make([]int, len(program))
			copy(program_p2, program)
			program_p2[1]=noun
			program_p2[2]=verb
			var content_p2 []int = int_code(program_p2)
			if content_p2[0]==19690720{
				fmt.Println("--- Part Two ---")
				fmt.Println(100*noun+verb)
			}
		}
	}
}

func int_code(code []int) []int{
	for i := 0; i+4 < len(code); i++ {
		if code[i]==1{
			code[code[i+3]] = code[code[i+1]]+code[code[i+2]]
			i+=3
		} else if code[i]==2{
			code[code[i+3]] = code[code[i+1]]*code[code[i+2]]
			i+=3
		}
		
	}
	return code
}
func readFile(filename string) string {
	bytes, err := ioutil.ReadFile(filename)
	check(err)
	return strings.TrimSpace(string(bytes))
}

func toInt(s string) int {
	result, err := strconv.Atoi(s)
	check(err)
	return result
}

func check(err error) {
	if err != nil {
		panic(err)
	}
}